app/quests/views.py

```diff
@@ -220,51 +220,66 @@ def index(request):
+            quest_package = get_package_helper(quest_qs, request)
+            package = ('Search', quest_package)
             quests.append(package)
-
-    print(f" phase1.3 at {round(time.time(),2)} ")
-    # new quests!
-    new_quests = Quest.objects.filter(visible=True, created_on__gt=(timezone.now() - timezone.timedelta(hours=hours_new))).order_by('-ui_data__success_pct')
-    if new_quests.exists():
-        quests.append(('New', get_package_helper(new_quests, request)))
-
-    print(f" phase1.4 at {round(time.time(),2)} ")
-    # popular quests
-    popular = Quest.objects.filter(visible=True).order_by('-ui_data__attempts_count')[0:5]
-    if popular.exists():
-        quests.append(('Popular', get_package_helper(popular, request)))
+        print(f" phase1.1 at {round(time.time(),2)} ")
+
+        # beaten/unbeaten
+        if request.user.is_authenticated:
+            attempts = request.user.profile.quest_attempts
+            if attempts.exists():
+                beaten = Quest.objects.filter(pk__in=attempts.filter(success=True).values_list('quest', flat=True))
+                unbeaten = Quest.objects.filter(pk__in=attempts.filter(success=False).exclude(quest__in=beaten).values_list('quest', flat=True))
+                if unbeaten.exists():
+                    quests.append(('Attempted', get_package_helper(unbeaten, request)))
+                    if selected_tab != 'Search':
+                        selected_tab = 'Attempted'
+                if beaten.exists():
+                    quests.append(('Beaten', get_package_helper(beaten, request)))
+                created_quests = request.user.profile.quests_created.filter(visible=True)
+                if created_quests:
+                    quests.append(('Created', get_package_helper(created_quests, request)))
+
+        print(f" phase1.2 at {round(time.time(),2)} ")
+        # difficulty tab
+        for diff in Quest.DIFFICULTIES:
+            quest_qs = Quest.objects.filter(difficulty=diff[0], visible=True).order_by('-ui_data__success_pct')
+            quest_package = get_package_helper(quest_qs, request)
+            package = (diff[0], quest_package)
+            if quest_qs.exists():
+                quests.append(package)
+
+        print(f" phase1.3 at {round(time.time(),2)} ")
+        # new quests!
+        new_quests = Quest.objects.filter(visible=True, created_on__gt=(timezone.now() - timezone.timedelta(hours=hours_new))).order_by('-ui_data__success_pct')
+        if new_quests.exists():
+            quests.append(('New', get_package_helper(new_quests, request)))
+
+        print(f" phase1.4 at {round(time.time(),2)} ")
+        # popular quests
+        popular = Quest.objects.filter(visible=True).order_by('-ui_data__attempts_count')[0:5]
```

Comment:
Since this queryset is being sliced, I think we can skip the `exists()` check. Actually, if there are less than 5 quests, I think this statement will fail with an index error

Reply:
Hmm I figured slicing would force queryset evaluation but apparently not for simple slicing, nevermind!